### PR TITLE
Only save partial chunk if chunk is completed

### PR DESCRIPTION
**Pull request by mzhangmzz**

### CODE REVIEW

**Comment by mzhangmzz** [comment]
> Can we test this? ^^

Yeah I think this case is already covered by existing CI tests. Which test broke when you added PartialChunk to INSERT_ONLY_COLS?

**Comment by matklad** [comment]
I think I was looking at chunks_produced_and_distributed_2_vals_per_shard.

But I am wondering if we can/should write a functional test here?

**Comment by matklad** [comment]
Verified that this indeed fixes the overwrite!

**Review by matklad — APPROVED** [review]
LGTM, on the "surface syntax" level, but I don't yet have the understanding of the undelying logic here.

File: chain/chunks/src/lib.rs [hunk 1]
```diff
@@ -1796,6 +1796,9 @@ impl ShardsManager {
         )?;
         let have_all_seal = seal.process(entry);*/
 
+        // If we don't care about the shard, we don't need to reconstruct the full chunk for
```
Comment (mm-near) [inline chain/chunks/src/lib.rs:1799]: I'd move this to 1810

File: chain/chunks/src/lib.rs [hunk 2]
```diff
@@ -1804,22 +1807,18 @@ impl ShardsManager {
...
-            // If we do care about the shard, we will remove the request once the full chunk is
-            //    assembled.
-            if !cares_about_shard {
                 self.complete_chunk(&chunk_hash);
                 return Ok(ProcessPartialEncodedChunkResult::HaveAllPartsAndReceipts);
             }
         }
 
+        // If we do care about the shard, we will remove the request once the full chunk is
+        //    assembled.
```
Comment (mm-near) [inline chain/chunks/src/lib.rs:1821]: so 'can_reconstruct' means that we also get get all the necessary receipts, rigth?

**Review by mm-near — APPROVED** [review]

**Comment by mzhangmzz** [comment]
> Can we test this? ^^

chunks_produced_and_distributed_2_vals_per_shard is actually a functional test for the logic touched in this PR.

The only test I think are not covered is when some nodes don't track all shards. We don't have that enabled now, I'll make a note to add test for this when chunk-only producers are enabled.

**Pull request merged**